### PR TITLE
calculate range comparision method handle only once per SortedRangeSet

### PR DESCRIPTION
**Pull request by kabunchi**
While using LargeDynamicFilters enabled which resulted in ±300K ranges SortedRangeSet we've got some lock contention in this backtrace when calling `SortedRangeSet.getOrderedRanges` from multiple threads :

`java.util.concurrent.ConcurrentHashMap.computeIfAbsent(java.base@11.0.12/ConcurrentHashMap.java:1723)
        - waiting to lock <0x00007f05c1c6fc60> (a java.util.concurrent.ConcurrentHashMap$Node)
        at io.trino.spi.type.TypeOperators.lambda$new$1(TypeOperators.java:72)
        at io.trino.spi.type.TypeOperators$$Lambda$4659/0x00007ede7d71fcb0.apply(Unknown Source)
        at io.trino.spi.type.TypeOperators.getOperatorAdaptor(TypeOperators.java:169)
        at io.trino.spi.type.TypeOperators.getOperatorAdaptor(TypeOperators.java:163)
        at io.trino.spi.type.TypeOperators.getComparisonUnorderedLastOperator(TypeOperators.java:125)
        at io.trino.spi.predicate.Range.<init>(Range.java:52)
        at io.trino.spi.predicate.SortedRangeSet$RangeView.toRange(SortedRangeSet.java:1132)
        at io.trino.spi.predicate.SortedRangeSet.getRange(SortedRangeSet.java:475)
        at io.trino.spi.predicate.SortedRangeSet.getOrderedRanges(SortedRangeSet.java:340)`
        
   when calling `getComparisonUnorderedLastOperator` once per `SortedRangeSet` the contention is reduced...

### CODE REVIEW

**Review by sopel39 — APPROVED** [review]

**Comment by kabunchi** [comment]
Changed only typo in commit message.

**Comment by martint** [comment]
Separately, we should look into why `computeIfAbsent` is causing so much contention. `TypeOperators` has an internal cache which should be uncontended when the entry is already present. This PR works around the local problem, but if there is an underlying issue with computeIfAbsent, it's going to affect other places.

**Comment by sopel39** [comment]
> Separately, we should look into why computeIfAbsent is causing so much contention

Right. @kabunchi which Java version do you use? Interesting related thread: https://stackoverflow.com/questions/26481796/concurrenthashmap-computeifabsent. Particularly https://stackoverflow.com/a/50193115

**Comment by sopel39** [comment]
@martint maybe in `TypeOperators` we should first call `get` and only then `computeIfAbsent`?

**Comment by sopel39** [comment]
Another thread: https://concurrency-interest.altair.cs.oswego.narkive.com/0Jfe1waD/computeifabsent-optimized-for-missing-entries

**Comment by kabunchi** [comment]
>

**Comment by kabunchi** [comment]
> @kabunchi which Java version do you use?

openjdk 11.0.10 2021-01-19 LTS

File: core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java [hunk 1]
```diff
@@ -83,6 +84,7 @@ private SortedRangeSet(Type type, boolean[] inclusive, Block sortedRanges)
         this.hashCodeOperator = TUPLE_DOMAIN_TYPE_OPERATORS.getHashCodeOperator(type, simpleConvention(FAIL_ON_NULL, BLOCK_POSITION));
         // choice of placing unordered values first or last does not matter for this code
         this.comparisonOperator = TUPLE_DOMAIN_TYPE_OPERATORS.getComparisonUnorderedLastOperator(type, simpleConvention(FAIL_ON_NULL, BLOCK_POSITION, BLOCK_POSITION));
+        this.rangeComparisonOperator = Range.getComparisonOperator(type);
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java:87]: Add code comment why we're doing that.

File: core/trino-spi/src/main/java/io/trino/spi/predicate/Range.java [hunk 2]
```diff
@@ -45,11 +45,14 @@
     private final boolean isSingleValue;
 
     Range(Type type, boolean lowInclusive, Optional<Object> lowValue, boolean highInclusive, Optional<Object> highValue)
+    {
+        this(type, lowInclusive, lowValue, highInclusive, highValue, getComparisonOperator(type));
+    }
+
+    Range(Type type, boolean lowInclusive, Optional<Object> lowValue, boolean highInclusive, Optional<Object> highValue, MethodHandle comparisonOperator)
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/predicate/Range.java:52]: Now that this is a constructor parameter, add `requireNonNull` check

**Comment by findepi** [comment]
>  This PR works around the local problem, but if there is an underlying issue with computeIfAbsent, it's going to affect other places.

That's true. However, pulling from the cache isn't going to be as fast as reading from a field, even in uncontended case, so this seems to me still a reasonable change.

**Comment by findepi** [comment]
> Particularly https://stackoverflow.com/a/50193115

And sibling one (https://stackoverflow.com/a/38362627).

What this seems to be saying to me is: stay clear from `ConcurrentHashMap.computeIfAbsent`.

**Comment by sopel39** [comment]
We can keep this PR, I've also created https://github.com/trinodb/trino/pull/10246

File: core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java (hunk 1 above)
Comment (kabunchi) [reply core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java:88]: done

File: core/trino-spi/src/main/java/io/trino/spi/predicate/Range.java (hunk 2 above)
Comment (kabunchi) [reply core/trino-spi/src/main/java/io/trino/spi/predicate/Range.java:52]: done

**Comment by martint** [comment]
> That's true. However, pulling from the cache isn't going to be as fast as reading from a field, even in uncontended case, so this seems to me still a reasonable change.

Agreed. I was just point out that we should look at the underlying problem as well, otherwise it will be an endless chase of similar problems in other parts of the code.

**Pull request merged**